File: src/core/validation.py
"""Server-side validation utilities for form inputs."""

import json
import logging
import re
from typing import Any
from urllib.parse import urlparse

logger = logging.getLogger(__name__)


class ValidationError(Exception):
    """Raised when validation fails."""

    def __init__(self, field: str, message: str):
        self.field = field
        self.message = message
        super().__init__(f"{field}: {message}")


class FormValidator:
    """Form validation utility class."""

    @staticmethod
    def validate_email(email: str) -> str | None:
        """Validate email address format."""
        if not email:
            return "Email is required"

        email_regex = r"^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$"
        if not re.match(email_regex, email):
            return "Invalid email address format"

        return None

    @staticmethod
    def validate_url(url: str, required: bool = False) -> str | None:
        """Validate URL format."""
        if not url and required:
            return "URL is required"

        if not url:
            return None

        try:
            result = urlparse(url)
            if not all([result.scheme, result.netloc]):
                return "Invalid URL format"

            if result.scheme not in ["http", "https"]:
                return "URL must use http or https protocol"

            return None
        except Exception as e:
            logger.debug(f"URL validation error: {e}")
            return "Invalid URL format"

    @staticmethod
    def validate_webhook_url(url: str) -> str | None:
        """Validate webhook URL (more specific than general URL)."""
        if not url:
            return None  # Webhooks are optional

        # First do general URL validation
        url_error = FormValidator.validate_url(url)
        if url_error:
            return url_error

        # Check for known webhook patterns
        if "hooks.slack.com/services/" in url:
            # Validate Slack webhook format
            parts = url.split("/")
            if len(parts) < 7:
                return "Invalid Slack webhook URL format"

        return None

    @staticmethod
    def validate_json(json_str: str, required: bool = True) -> str | None:
        """Validate JSON string."""
        if not json_str and required:
            return "JSON configuration is required"

        if not json_str:
            return None

        try:
            json.loads(json_str)
            return None
        except json.JSONDecodeError as e:
            return f"Invalid JSON: {str(e)}"

    @staticmethod
    def validate_principal_id(principal_id: str) -> str | None:
        """Validate principal ID format."""
        if not principal_id:
            return "Principal ID is required"

        if not re.match(r"^[a-zA-Z0-9_-]+$", principal_id):
            return "Principal ID can only contain letters, numbers, underscores, and hyphens"

        if len(principal_id) < 3:
            return "Principal ID must be at least 3 characters long"

        if len(principal_id) > 50:
            return "Principal ID must be less than 50 characters"

        return None

    @staticmethod
    def validate_network_id(network_id: str) -> str | None:
        """Validate network ID (should be numeric)."""
        if not network_id:
            return "Network ID is required"

        if not network_id.isdigit():
            return "Network ID must be numeric"

        return None

    @staticmethod
    def validate_required(value: str, field_name: str = "Field") -> str | None:
        """Validate that a field is not empty."""
        if not value or not value.strip():
            return f"{field_name} is required"
        return None

    @staticmethod
    def validate_length(
        value: str, min_length: int | None = None, max_length: int | None = None, field_name: str = "Field"
    ) -> str | None:
        """Validate string length."""
        if not value:
            return None

        if min_length and len(value) < min_length:
            return f"{field_name} must be at least {min_length} characters"

        if max_length and len(value) > max_length:
            return f"{field_name} must be less than {max_length} characters"

        return None

    @staticmethod
    def validate_tenant_name(name: str) -> str | None:
        """Validate tenant name."""
        if error := FormValidator.validate_required(name, "Tenant name"):
            return error

        if error := FormValidator.validate_length(name, min_length=3, max_length=100, field_name="Tenant name"):
            return error

        return None

    @staticmethod
    def validate_subdomain(subdomain: str) -> str | None:
        """Validate subdomain format."""
        if not subdomain:
            return "Subdomain is required"

        # Allow localhost for development
        if subdomain == "localhost":
            return None

        if not re.match(r"^[a-z0-9-]+$", subdomain):
            return "Subdomain can only contain lowercase letters, numbers, and hyphens"

        if subdomain.startswith("-") or subdomain.endswith("-"):
            return "Subdomain cannot start or end with a hyphen"

        if len(subdomain) < 3:
            return "Subdomain must be at least 3 characters long"

        if len(subdomain) > 63:
            return "Subdomain must be less than 63 characters"

        return None

    @staticmethod
    def validate_role(role: str) -> str | None:
        """Validate user role."""
        valid_roles = ["admin", "manager", "viewer"]
        if role not in valid_roles:
            return f"Invalid role. Must be one of: {', '.join(valid_roles)}"
        return None


def validate_form_data(data: dict[str, Any], validators: dict[str, list] | list[str]) -> tuple[bool, list[str]]:
    """
    Validate form data using specified validators.

    Args:
        data: Form data dictionary
        validators: Either a dictionary mapping field names to list of validator functions,
                   or a list of required field names for simple presence validation

    Returns:
        Tuple of (is_valid, list of error messages)
    """
    errors: list[str] = []

    # Handle simple required field validation when passed a list
    if isinstance(validators, list):
        for field in validators:
            if not data.get(field, "").strip():
                errors.append(f"{field.title()} is required")
        return (len(errors) == 0, errors)

    # Handle dictionary of validators
    for field, field_validators in validators.items():
        value = data.get(field, "")

        for validator in field_validators:
            if not callable(validator):
                continue

            error = validator(value)
            if error:
                errors.append(f"{field.title()}: {error}")
                break  # Stop on first error for this field

    return (len(errors) == 0, errors)


def sanitize_json(json_str: str) -> str:
    """Sanitize and format JSON string."""
    try:
        # Parse and re-serialize to ensure valid JSON
        parsed = json.loads(json_str)
        return json.dumps(parsed, indent=2)
    except json.JSONDecodeError:
        return json_str  # Return as-is if not valid JSON


def sanitize_url(url: str) -> str:
    """Sanitize URL by ensuring proper format."""
    if not url:
        return url

    # Ensure URL has a scheme
    if not url.startswith(("http://", "https://")):
        url = "https://" + url

    # Remove trailing slashes
    return url.rstrip("/")


def normalize_agent_url(url: str) -> str:
    """Normalize agent URL to base form for consistent comparison.

    Strips common path suffixes that users might include:
    - /mcp
    - /a2a
    - /.well-known/adcp/sales
    - Trailing slashes

    This ensures all variations of an agent URL normalize to the same base URL:
        "https://creative.adcontextprotocol.org/" -> "https://creative.adcontextprotocol.org"
        "https://creative.adcontextprotocol.org/mcp" -> "https://creative.adcontextprotocol.org"
        "https://creative.adcontextprotocol.org/a2a" -> "https://creative.adcontextprotocol.org"
        "https://publisher.com/.well-known/adcp/sales" -> "https://publisher.com"

    Args:
        url: Agent URL to normalize

    Returns:
        Normalized base URL
    """
    if not url:
        return url

    # First, remove trailing slashes
    normalized = url.rstrip("/")

    # Common path suffixes to strip (order matters - longest first)
    suffixes_to_strip = [
        "/.well-known/adcp/sales",
        "/mcp",
        "/a2a",
    ]

    # Strip each suffix (check multiple times in case of multiple trailing slashes)
    for suffix in suffixes_to_strip:
        if normalized.endswith(suffix):
            normalized = normalized[: -len(suffix)]
            # Remove any trailing slashes that remain
            normalized = normalized.rstrip("/")
            break  # Only strip one suffix

    return normalized


def sanitize_form_data(data: dict[str, Any]) -> dict[str, Any]:
    """Sanitize form data before saving."""
    sanitized = {}

    for key, value in data.items():
        if isinstance(value, str):
            # Trim whitespace
            value = value.strip()

            # Sanitize specific field types
            if "url" in key.lower():
                value = sanitize_url(value)
            elif key == "config" or "json" in key.lower():
                value = sanitize_json(value)

        sanitized[key] = value

    return sanitized


# GAM-specific validation functions
def validate_gam_network_code(network_code: str) -> str | None:
    """Validate GAM network code format."""
    if not network_code:
        return None  # Optional field

    # Network codes should be numeric and reasonable length
    if not re.match(r"^\d{1,20}$", network_code):
        return "Network code must be numeric and up to 20 digits"

    return None


def validate_gam_trafficker_id(trafficker_id: str) -> str | None:
    """Validate GAM trafficker ID format."""
    if not trafficker_id:
        return None  # Optional field

    # Trafficker IDs should be numeric and reasonable length
    if not re.match(r"^\d{1,20}$", trafficker_id):
        return "Trafficker ID must be numeric and up to 20 digits"

    return None


def validate_gam_refresh_token(refresh_token: str) -> str | None:
    """Validate GAM refresh token format and length."""
    if not refresh_token:
        return "Refresh token is required"

    # Basic length validation (refresh tokens are typically long)
    if len(refresh_token) < 20:
        return "Refresh token appears to be invalid (too short)"

    if len(refresh_token) > 1000:
        return "Refresh token is too long (max 1000 characters)"

    # Check for common invalid patterns
    if refresh_token.startswith("Bearer "):
        return "Do not include 'Bearer ' prefix in refresh token"

    return None


def validate_gam_config(data: dict[str, Any]) -> dict[str, str | None]:
    """Validate all GAM configuration fields."""
    errors: dict[str, str | None] = {}

    # Validate network code
    if "network_code" in data:
        error = validate_gam_network_code(str(data["network_code"]))
        if error:
            errors["network_code"] = error

    # Validate trafficker ID
    if "trafficker_id" in data:
        error = validate_gam_trafficker_id(str(data["trafficker_id"]))
        if error:
            errors["trafficker_id"] = error

    # Validate refresh token
    if "refresh_token" in data:
        error = validate_gam_refresh_token(data["refresh_token"])
        if error:
            errors["refresh_token"] = error

    return errors

File: static/validation.js
// Form validation utilities for AdCP Admin UI

// Validation rules
const validators = {
    // Email validation
    email: (value) => {
        const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
        if (!emailRegex.test(value)) {
            return 'Please enter a valid email address';
        }
        return null;
    },

    // URL validation (for webhooks)
    url: (value) => {
        if (!value) return null; // Optional
        try {
            const url = new URL(value);
            if (!['http:', 'https:'].includes(url.protocol)) {
                return 'URL must start with http:// or https://';
            }
            return null;
        } catch (e) {
            return 'Please enter a valid URL';
        }
    },

    // Webhook URL validation (more specific)
    webhookUrl: (value) => {
        if (!value) return null; // Optional
        const urlError = validators.url(value);
        if (urlError) return urlError;

        // Check for common webhook patterns
        if (value.includes('hooks.slack.com/services/')) {
            // Slack webhook validation
            const parts = value.split('/');
            if (parts.length < 7) {
                return 'Invalid Slack webhook URL format';
            }
        }
        return null;
    },

    // JSON validation
    json: (value) => {
        try {
            JSON.parse(value);
            return null;
        } catch (e) {
            return `Invalid JSON: ${e.message}`;
        }
    },

    // Principal ID validation
    principalId: (value) => {
        if (!value) return 'Principal ID is required';
        if (!/^[a-zA-Z0-9_-]+$/.test(value)) {
            return 'Principal ID can only contain letters, numbers, underscores, and hyphens';
        }
        if (value.length < 3) {
            return 'Principal ID must be at least 3 characters long';
        }
        if (value.length > 50) {
            return 'Principal ID must be less than 50 characters';
        }
        return null;
    },

    // Network ID validation (numeric)
    networkId: (value) => {
        if (!value) return 'Network ID is required';
        if (!/^\d+$/.test(value)) {
            return 'Network ID must be numeric';
        }
        return null;
    },

    // Required field validation
    required: (value) => {
        if (!value || value.trim() === '') {
            return 'This field is required';
        }
        return null;
    },

    // Min length validation
    minLength: (min) => (value) => {
        if (value && value.length < min) {
            return `Must be at least ${min} characters`;
        }
        return null;
    },

    // Max length validation
    maxLength: (max) => (value) => {
        if (value && value.length > max) {
            return `Must be less than ${max} characters`;
        }
        return null;
    }
};

// Show error message for a field
function showError(fieldId, message) {
    const field = document.getElementById(fieldId);
    if (!field) return;

    // Remove existing error
    clearError(fieldId);

    // Add error class
    field.classList.add('error');

    // Create error message element
    const errorDiv = document.createElement('div');
    errorDiv.className = 'field-error';
    errorDiv.textContent = message;
    errorDiv.id = `${fieldId}-error`;

    // Insert after the field
    field.parentNode.insertBefore(errorDiv, field.nextSibling);
}

// Clear error message for a field
function clearError(fieldId) {
    const field = document.getElementById(fieldId);
    if (!field) return;

    field.classList.remove('error');

    const errorDiv = document.getElementById(`${fieldId}-error`);
    if (errorDiv) {
        errorDiv.remove();
    }
}

// Validate a single field
function validateField(fieldId, validatorFuncs) {
    const field = document.getElementById(fieldId);
    if (!field) return true;

    const value = field.value;

    for (const validator of validatorFuncs) {
        const error = validator(value);
        if (error) {
            showError(fieldId, error);
            return false;
        }
    }

    clearError(fieldId);
    return true;
}

// Validate entire form
function validateForm(formId, fieldValidators) {
    let isValid = true;

    for (const [fieldId, validators] of Object.entries(fieldValidators)) {
        if (!validateField(fieldId, validators)) {
            isValid = false;
        }
    }

    return isValid;
}

// Add real-time validation to a field
function addRealtimeValidation(fieldId, validatorFuncs) {
    const field = document.getElementById(fieldId);
    if (!field) return;

    // Validate on blur
    field.addEventListener('blur', () => {
        validateField(fieldId, validatorFuncs);
    });

    // Clear error on input (gives immediate feedback when user fixes error)
    field.addEventListener('input', () => {
        const hasError = field.classList.contains('error');
        if (hasError) {
            validateField(fieldId, validatorFuncs);
        }
    });
}

// Initialize validation for Slack form
function initSlackFormValidation() {
    const form = document.getElementById('slackForm');
    if (!form) return;

    // Add real-time validation
    addRealtimeValidation('slack_webhook_url', [validators.webhookUrl]);
    addRealtimeValidation('slack_audit_webhook_url', [validators.webhookUrl]);

    // Validate on submit
    form.addEventListener('submit', (e) => {
        const isValid = validateForm('slackForm', {
            'slack_webhook_url': [validators.webhookUrl],
            'slack_audit_webhook_url': [validators.webhookUrl]
        });

        if (!isValid) {
            e.preventDefault();
        }
    });
}

// Initialize validation for JSON config form
function initConfigFormValidation() {
    const configField = document.getElementById('config');
    if (!configField) return;

    // Add real-time validation
    addRealtimeValidation('config', [validators.required, validators.json]);

    // Format JSON on blur
    configField.addEventListener('blur', () => {
        try {
            const parsed = JSON.parse(configField.value);
            configField.value = JSON.stringify(parsed, null, 2);
        } catch (e) {
            // Invalid JSON, validation will handle the error
        }
    });
}

// Initialize validation for principal form
function initPrincipalFormValidation() {
    const form = document.querySelector('form[action*="/principals/create"]');
    if (!form) return;

    // Add real-time validation
    addRealtimeValidation('principal_id', [validators.required, validators.principalId]);
    addRealtimeValidation('name', [
        validators.required,
        validators.minLength(3),
        validators.maxLength(100)
    ]);

    // Validate on submit
    form.addEventListener('submit', (e) => {
        const isValid = validateForm('principalForm', {
            'principal_id': [validators.required, validators.principalId],
            'name': [validators.required, validators.minLength(3), validators.maxLength(100)]
        });

        if (!isValid) {
            e.preventDefault();
        }
    });
}

// Initialize validation for adapter setup forms
function initAdapterFormValidation() {
    // GAM validation
    addRealtimeValidation('gam_network_id', [validators.required, validators.networkId]);
    addRealtimeValidation('gam_credentials', [validators.required, validators.json]);

    // Kevel validation
    addRealtimeValidation('kevel_network_id', [validators.required, validators.networkId]);
    addRealtimeValidation('kevel_api_key', [validators.required, validators.minLength(10)]);

    // Triton validation
    addRealtimeValidation('triton_station_id', [validators.required, validators.minLength(3)]);
    addRealtimeValidation('triton_api_key', [validators.required, validators.minLength(10)]);
}

// Initialize validation for user form
function initUserFormValidation() {
    const form = document.querySelector('.add-user-form');
    if (!form) return;

    // Add email validation
    const emailInput = form.querySelector('input[type="email"]');
    if (emailInput) {
        addRealtimeValidation(emailInput.id, [validators.required, validators.email]);
    }
}

// Initialize all validations when DOM is loaded
document.addEventListener('DOMContentLoaded', () => {
    initSlackFormValidation();
    initConfigFormValidation();
    initPrincipalFormValidation();
    initAdapterFormValidation();
    initUserFormValidation();
});

// Export for use in other scripts
window.FormValidation = {
    validators,
    validateField,
    validateForm,
    showError,
    clearError,
    addRealtimeValidation
};
